internal/mithril-metric/src/commons.rs

```diff
@@ -0,0 +1,494 @@
+                            stringify!($metric_attribute)
+                        } else {
+                            $name
+                        };
+                        let $metric_attribute = $metric_type::new(
+                            logger.clone(),
+                            &format!("{}{}", prefix, metric_name),
+                            $help,
+                        )?;
+                        registry.register($metric_attribute.collector())?;
+                    )*
+
+                    Ok(Self {
+                        registry,
+                        $(
+                            $metric_attribute,
+                        )*
+                    })
+                }
+                $(
+                    /// Get the `$metric_attribute` counter.
+                    pub fn [<get_ $metric_attribute>](&self) -> &$metric_type {
+                        &self.$metric_attribute
+                    }
+                )*
+            }
+
+            impl MetricsServiceExporter for $service {
+                fn export_metrics(&self) -> StdResult<String> {
+                    metrics_tools::export_metrics(&self.registry)
+                }
+            }
+
+        }
+    };
+}
+
+/// Mithril metric
+pub trait MithrilMetric {
+    /// Metric name
+    fn name(&self) -> String;
+
+    /// Wrapped prometheus collector
+    fn collector(&self) -> Box<dyn Collector>;
+}
+
+pub struct MetricCounter {
+    name: String,
+    logger: Logger,
+    counter: Box<Counter>,
+}
+
+impl MetricCounter {
+    pub fn new(logger: Logger, name: &str, help: &str) -> StdResult<Self> {
+        let counter = MetricCounter::create_metric_counter(name, help)?;
+        Ok(Self {
+            logger,
+            name: name.to_string(),
+            counter: Box::new(counter),
+        })
+    }
+
+    pub fn record(&self) {
+        debug!(self.logger, "incrementing '{}' counter", self.name);
+        self.counter.inc();
+    }
+
+    pub fn get(&self) -> CounterValue {
+        self.counter.get().round() as CounterValue
+    }
+
+    fn create_metric_counter(name: &MetricName, help: &str) -> StdResult<Counter> {
+        let counter_opts = Opts::new(name, help);
+        let counter = Counter::with_opts(counter_opts)?;
+
+        Ok(counter)
+    }
+}
+
+impl MithrilMetric for MetricCounter {
+    fn collector(&self) -> Box<dyn Collector> {
+        self.counter.clone()
+    }
+
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+}
+
+pub struct MetricGauge {
+    name: String,
+    logger: Logger,
+    gauge: Box<Gauge>,
+}
+
+impl MetricGauge {
+    pub fn new(logger: Logger, name: &str, help: &str) -> StdResult<Self> {
+        let gauge = MetricGauge::create_metric_gauge(name, help)?;
+        Ok(Self {
+            logger,
+            name: name.to_string(),
+            gauge: Box::new(gauge),
+        })
+    }
+
+    pub fn record(&self, epoch: Epoch) {
```

Comment:
```suggestion
    pub fn record::<T: Into<f64>>(&self, value: T) {
```

Reply:
With this generic, when we want to record an epoch we need to dereference to u64 and cast the value
`my_gauge.record(*value as f64);`

We probably also want to return a f64 from `get()` function. In that case, the caller need to cast the result and then build an Epoch: `Epoch(my_gauge.get() as u64)`

Is it ok for you ?